atomics/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,145 @@
+    T return_val = *dest;
+    *dest = value;
+    // device_atomic_thread_fence(MemoryOrderRelease(), scope);
+    // unlock_address_openacc((void*)dest, scope);
+    return return_val;
+  }
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+T device_atomic_compare_exchange(
+    T* dest, T compare, T value, MemoryOrder, MemoryScope scope) {
+  if constexpr (std::is_integral_v<T> && ((sizeof(T) == 4) || (sizeof(T) == 8))) {
+    static_assert(sizeof(unsigned int) == 4);
+    static_assert(sizeof(unsigned long long int) == 8);
+    using cas_t =
+        std::conditional_t<(sizeof(T) == 4), unsigned int, unsigned long long int>;
+    cas_t return_val = atomicCAS(reinterpret_cast<cas_t*>(dest),
+                                 reinterpret_cast<cas_t&>(compare),
+                                 reinterpret_cast<cas_t&>(value));
+    return reinterpret_cast<T&>(return_val);
+#ifdef DESUL_CUDA_ARCH_IS_PRE_PASCAL
+  } else if constexpr (std::is_same_v<T, float>) {
+#else
+  } else if constexpr (std::is_same_v<T, float> || std::is_same_v<T, double>) {
+#endif
+    return atomicCAS(dest, compare, value);
+  } else {
+    // FIXME_OPENACC
+    printf(
+        "DESUL error in device_atomic_compare_exchange(): Not supported atomic "
+        "operation in the OpenACC backend\n");
+    T current_val = *dest;
+    // Acquire a lock for the address
+    // while (!lock_address_openacc((void*)dest, scope)) {
+    //}
+    // device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+    if (current_val == compare) {
+      *dest = value;
+      // device_atomic_thread_fence(MemoryOrderRelease(), scope);
+    }
+    // unlock_address_openacc((void*)dest, scope);
+    return current_val;
+  }
+}
+
+#else  // not NVHPC
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+T device_atomic_exchange(T* dest, T value, MemoryOrder, MemoryScope) {
+  if constexpr (std::is_arithmetic_v<T>) {
+    T return_val;
+#pragma acc atomic capture
+    {
+      return_val = *dest;
+      *dest = value;
+    }
+    return return_val;
+  } else {
+    // FIXME_OPENACC
+    printf(
+        "DESUL error in device_atomic_exchange(): Not supported atomic operation in "
+        "the OpenACC backend\n");
+    //  Acquire a lock for the address
+    // while (!lock_address_openacc((void*)dest, scope)) {
+    // }
+    // device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+    T return_val = *dest;
+    *dest = value;
+    // device_atomic_thread_fence(MemoryOrderRelease(), scope);
+    // unlock_address_openacc((void*)dest, scope);
+    return return_val;
+  }
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+T device_atomic_compare_exchange(
+    T* dest, T compare, T value, MemoryOrder, MemoryScope scope) {
+  // FIXME_OPENACC
+  printf(
+      "DESUL error in device_atomic_compare_exchange(): Not supported atomic operation "
+      "in the OpenACC backend\n");
```

Comment:
Given the issue with device_atomic_thread_fence being unimplementable, I get having a warning but why a printf, and why just this one rather than all of them when using a non-relaxed atomic?

Reply:
What else than printf could we do?  As far as I know we cannot abort from the device with OpenACC
https://github.com/kokkos/kokkos/blob/0e5aa150355cc5163198ae1cbba7d89fef2ab61f/core/src/Kokkos_Abort.hpp#L86-L87

This is the generic fallback so the warning will be raised if there is no overload that is more specialized.

Reply:
The current OpenACC atomic implementation is tested only by the NVHPC compiler (and this PR includes implementations only for the NVHPC compiler), but the general Kokkos OpenACC backend is supposed to support both the NVHPC compiler and Clacc/LLVM compiler.
After this PR is merged, I will implement the specialized versions for the Clacc OpenACC compiler.
(Clacc OpenACC compiler is built on OpenMP, and OpenMP supports both atomic_compare_exchange() and atomic_thread_fence().)

Reply:
Ok, that works for me then.  I'll take a pass over it but IIRC this looked good the last time I checked.